### PR TITLE
Wildcard field - add normalizer support

### PR DESCRIPTION
**Pull request by markharwood**
Add normalizer support for the wildcard field to allow case insensitivity

### CODE REVIEW

File: x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java [hunk 1]
```diff
@@ -327,6 +376,12 @@ public boolean equals(Object obj) {
 
         @Override
         public Query wildcardQuery(String wildcardPattern, RewriteMethod method, QueryShardContext context) {
+            try {
+                wildcardPattern = normalize(wildcardPattern);
```
Comment (jimczi) [inline x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java]: We should normalize all term-based query like ` KeywordFieldType`.
Also @cbuescher added the support in [`StringFieldType`](https://github.com/elastic/elasticsearch/pull/53512) to mask special characters (`*`)  during normalization so we should reuse here ?

**Review by jimczi — COMMENTED** [review]
I left one comment but adding the support for this option makes sense to me.

File: x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java [hunk 2]
```diff
@@ -467,7 +526,63 @@ public Query termsQuery(List<?> values, QueryShardContext context) {
...
+                    ts.end();
+                    return newValue;
+                }
+            }
+            return value;
+        }       
+        
+        private static final Pattern WILDCARD_PATTERN = Pattern.compile("(\\\\.)|([?*]+)");
+        
+        String normalizeWildcardPattern(String value) throws IOException {
```
Comment (jimczi) [inline x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java]: Can we try to expose something reusable (a static function ?) in `StringFieldType` ?

File: x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java [hunk 3]
```diff
@@ -206,13 +235,37 @@ public WildcardFieldType() {
...
+        }       
+        
+        private NamedAnalyzer normalizer() {
+            return normalizer;
+        }
+
+        public void setNormalizer(NamedAnalyzer normalizer) {
+            checkIfFrozen();
+            this.normalizer = normalizer;
+        }
```
Comment (jimczi) [inline x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java]: You also need to override `checkCompatibility` in order to ensure that the `normalizer` doesn't change ?

**Review by jimczi — COMMENTED** [review]
I left 2 more comments

**Comment by markharwood** [comment]
Good to go, @jimczi ?

**Review by jimczi — APPROVED** [review]
LGTM

**Comment by markharwood** [comment]
Backport to 7x pending on https://github.com/elastic/elasticsearch/pull/53512

**Pull request merged**